news/inn/doc/qio.3
.\" $Revision: 1.1 $
.TH QIO 3
.SH NAME
qio \- quick I/O part of InterNetNews library
.SH SYNOPSIS
.nf
.ta \w'    unsigned long    'u
.B
#include "qio.h"

.B "QIOSTATE *"
.B "QIOopen(name, size)"
.B "    char	*name;"
.B "    int	size;"

.B "QIOSTATE *"
.B "QIOfdopen(fd, size)"
.B "    int	fd;"
.B "    int	size;"

.B "void"
.B "QIOclose(qp)"
.B "    QIOSTATE	*qp;"

.B "char *"
.B "QIOread(qp)"
.B "    QIOSTATE	*qp;"

.B "int"
.B "QIOlength(qp)"
.B "    QIOSTATE	*qp;"

.B "int"
.B "QIOtoolong(qp)"
.B "    QIOSTATE	*qp;"

.B "int"
.B "QIOerror(qp)"
.B "    QIOSTATE	*qp;"

.B "int"
.B "QIOtell(qp)"
.B "    QIOSTATE	*qp;"

.B "int"
.B "QIOrewind(qp)"
.B "    QIOSTATE	*qp;"

.B "int"
.B "QIOfileno(qp)"
.B "    QIOSTATE	*qp;"
.SH DESCRIPTION
The routines described in this manual page are part of the InterNetNews
library,
.IR libinn (3).
They are used to provide quick read access to files.
The letters ``QIO'' stand for
.IR Q uick
.IR I/O .
.PP
.I QIOopen
opens the file
.I name
for reading.
It uses a buffer of
.I size
bytes, which must also be larger then the longest expected line.
The header file defines the constant QIO_BUFFER as a reasonable default.
If
.I size
is zero, then
.\" =()<.if '@<HAVE_ST_BLKSIZE>@'DO' \{\>()=
.if 'DO'DO' \{\
.I QIOopen
will call
.IR stat (2)
and use the returned block size; if that fails\}
it will use QIO_BUFFER.
It returns NULL on error, or a pointer to a handle to be used in other calls.
.I QIOfdopen
performs the same function except that
.I fd
refers to an already-open descriptor.
.PP
.I QIOclose
closes the open file and releases any resources used by it.
.PP
.I QIOread
returns a pointer to the next line in the file.
The trailing newline will be replaced with a \e0.
If EOF is reached, an error occurs, or if the line is longer than the
buffer,
.I QIOread
returns NULL.
.PP
After a successful call to
.IR QIOread ,
.I QIOlength
will return the length of the current line.
.PP
The functions
.I QIOtoolong
and
.I QIOerror
can be called after
.I QIOread
returns NULL to determine if there was an error, or if the line was
too long.
If
.I QIOtoolong
returns non-zero, then the current line did not fit in the buffer, and the
next call to
.I QIOread
will try read the rest of the line.
Long lines can only be discarded.
If
.I QIOerror
returns non-zero, then a serious I/O error occurred.
.PP
.I QIOtell
returns the
.IR lseek (2)
offset at which the next line will start.
.PP
.I QIOrewind
sets the read pointer back to the beginning of the file.
.PP
.I QIOfileno
returns the descriptor of the open file.
.PP
.IR QIOlength ,
.IR QIOtoolong ,
.IR QIOerror ,
.IR QIOtell ,
and
.I QIOfileno
are implemented as macro's defined in the header file.
.SH EXAMPLE
.RS
.nf
QIOSTATE	*h;
long	offset;
char	*p;

h = QIOopen("/etc/motd", QIO_BUFFER);
for (offset = QIOtell(h); (p = QIOread(h)) != NULL; offset = QIOtell(h))
    printf("At %ld, %s\en", offset, p);
if (QIOerror(h)) {
    perror("Read error");
    exit(1);
}
QIOclose(h);
.fi
.RE
.SH HISTORY
Written by Rich $alz <rsalz@uunet.uu.net> for InterNetNews.
.de R$
This is revision \\$3, dated \\$4.
..
.R$ $Id: qio.3,v 1.1 1993/08/27 02:46:07 alm Exp $


news/inn/lib/qio.c
/*  $Revision: 1.1 $
**
**  Quick I/O package -- optimized for reading through a file.
*/
#include "configdata.h"
#include <stdio.h>
#include <fcntl.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <errno.h>
#include "libinn.h"
#include "clibrary.h"
#include "qio.h"
#include "macros.h"


/*
**  Open a quick file from a descriptor.
*/
QIOSTATE *
QIOfdopen(fd, size)
    int		fd;
    int		size;
{
    QIOSTATE	*qp;
#if	defined(DO_HAVE_ST_BLKSIZE)
    struct stat	Sb;
#endif	/* defined(DO_HAVE_ST_BLKSIZE) */

    qp = NEW(QIOSTATE, 1);
    qp->flag = QIO_ok;
    qp->fd = fd;
#if	defined(DO_HAVE_ST_BLKSIZE)
    if (size == 0)
	size = fstat(fd, &Sb) >= 0 ? (int)Sb.st_blksize : QIO_BUFFER;
#else
    if (size == 0)
	size = QIO_BUFFER;
#endif	/* defined(DO_HAVE_ST_BLKSIZE) */
    qp->Size = size;
    qp->Buffer = NEW(char, size);
    qp->Count = 0;
    qp->Start = qp->Buffer;
    qp->End = qp->Buffer;

    return qp;
}


/*
**  Open a file for reading.
*/
QIOSTATE *
QIOopen(name, size)
    char	*name;
    int		size;
{
    int		fd;

    /* Open the file, read in the first chunk. */
    if ((fd = open(name, O_RDONLY)) < 0)
	return NULL;
    return QIOfdopen(fd, size);
}


/*
**  Close an open stream.
*/
void
QIOclose(qp)
    QIOSTATE	*qp;
{
    (void)close(qp->fd);
    DISPOSE(qp->Buffer);
    DISPOSE(qp);
}


/*
**  Rewind an open stream.
*/
int
QIOrewind(qp)
    QIOSTATE	*qp;
{
    int		i;

    if (lseek(qp->fd, 0L, SEEK_SET) == -1)
	return -1;
    i = read(qp->fd, qp->Buffer, (SIZE_T)qp->Size);
    if (i < 0)
	return i;
    qp->Count = i;
    qp->Start = qp->Buffer;
    qp->End = &qp->Buffer[i];
    return 0;
}


/*
**  Get the next line from the input.
*/
char *
QIOread(qp)
    QIOSTATE	*qp;
{
    register char	*p;
    register char	*q;
    char		*save;
    int			i;

    /* Read from buffer if there is any data there. */
    if (qp->End > qp->Start) {

	/* Find the newline. */
	p = memchr((POINTER)qp->Start, '\n', (SIZE_T)(qp->End - qp->Start));
	if (p != NULL) {
	    *p = '\0';
	    qp->Length = p - qp->Start;
	    save = qp->Start;
	    qp->Start = p + 1;
	    qp->flag = QIO_ok;
	    return save;
	}

	/* Not there; move unread part down to start of buffer. */
	for (p = qp->Buffer, q = qp->Start; q < qp->End; )
	    *p++ = *q++;
    }
    else
	p = qp->Buffer;

    /* Read data, reset the pointers. */
    i = read(qp->fd, p, (SIZE_T)(&qp->Buffer[qp->Size] - p));
    if (i < 0) {
	qp->flag = QIO_error;
	return NULL;
    }
    if (i == 0) {
	qp->flag = QIO_ok;
	return NULL;
    }
    qp->Count += i;
    qp->Start = qp->Buffer;
    qp->End = &p[i];

    /* Now try to find it. */
    p = memchr((POINTER)qp->Start, '\n', (SIZE_T)(qp->End - qp->Start));
    if (p != NULL) {
	*p = '\0';
	qp->Length = p - qp->Start;
	save = qp->Start;
	qp->Start = p + 1;
	qp->flag = QIO_ok;
	return save;
    }

    /* Still not there -- line is too long. */
    qp->flag = QIO_long;
    qp->Start = qp->End;
    return NULL;
}



#if	defined(TEST)
int
main(ac, av)
    int		ac;
    char	*av[];
{
    QIOSTATE	*h;
    char	*p;
    long	where;

    h = QIOopen(ac > 1 ? av[1] : "/usr/lib/news/history", 0);
    if (h == NULL) {
	perror("Can't open file");
	exit(1);
    }

    where = QIOtell(h);
    while ((p = QIOread(h)) != NULL) {
	printf("%5ld %3d %s\n", where, QIOlength(h), p);
	where = QIOtell(h);
    }
    if (QIOtoolong(h)) {
	fprintf(stderr, "Line too line at %ld\n", QIOtell(h));
	exit(1);
    }
    if (QIOerror(h)) {
	perror("Can't read");
	exit(1);
    }
    QIOclose(h);
    exit(0);
    /* NOTREACHED */
}
#endif	/* defined(TEST) */
